### ISSUE TITLE
Preferred net/http.Client for storage drivers?

### ISSUE THREAD

**Issue by ahmetb** [distribution/distribution#1880]
Hi folks, I am wondering if we have a common practice around net/http.Client values in storage drivers. I remember seeing some issues in Azure driver hanging while pushing images and I experienced some of that myself. This [cloudflare blog on http client timeouts](https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/) shows that the default values regarding timeouts/deadlines on the `http.DefaultClient` are not very good and suitable for public-Internet access.

For the Azure driver, we have a way of swapping out the net/http.Client instance used. Perhaps if we can get an instance from the registry somehow, we could use that? I'd love to see what you folks think about having a common practice for http communications from registry to Internet endpoints.

**Comment by RichardScothern** [comment]
hi @ahmetalpbalkan 

That is an interesting article.

I had a quick look through the codebase:

For the registry itself, we do use `ListenAndServe()` which looks like it could be a problem.  The registry HTTP client has a 60s timeout (for dial, handshake, request & response) which the docker engine uses.  This represents the majority of registry clients which is why I think we don't see dangling FDs.  @nwt does this sound right?

The notification HTTP client has a configurable timeout.

The swift driver client has a 60s connection timeout and a 15 minute(!) total timeout.

Everything else looks like it defaults to `http.DefaultClient`

In summary: we are not consistent with out timeouts.  I don't think replacing the client wholesale would work (we would override any previously set settings) in all drivers, but a common set of timeouts that we set makes sense.

Do you have a handle on exactly why the Azure driver is hanging?

**Comment by ahmetalpbalkan** [comment]
> Do you have a handle on exactly why the Azure driver is hanging?

not really, it seems to be a sporadic issue. It sometimes even hangs on my Raspberry PI with Azure Python SDK so I bet it might be a backend problem on the Azure side.

However, I think if I ever hit this issue myself or hear someone hitting it, I will customize the http client on the azure driver.

Just wanted to bring this to your attention, I thought it might be good to set a general practice around the external outbound connectivity.

**Comment by milosgajdos** [comment]
Closing as wildly outdated. Besides, we've moved the existing API client provided by this codebase into `internal` package and explicitly recommend folks not to use it. The goal of this project is not to provide clients but rather the APIs. Feel free to move this question to GH discussions or Slack.
